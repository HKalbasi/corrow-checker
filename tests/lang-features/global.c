#include <stddef.h>

void *owned malloc(unsigned long size);
void free(void *owned ptr);

void* global_config;

enum Foo {
    Foo1 = 1,
    Foo2 = 2,
};

int main()
{
    void* p1 = malloc(Foo1);
    void* p2 = malloc(Foo2);
    global_config = p1;
    global_config = p2;
}
