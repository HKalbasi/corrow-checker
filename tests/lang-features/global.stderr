Error: Potential leak detected
    ╭─[preproccessed_file.c:1:1]
    │
 38 │     void* p1 = malloc(Foo1);
    │              ───────┬┬─────  
    │                     ╰──────── Ownership of the value transferred to this place here
    │                      │       
    │                      ╰─────── Ownership of the value transferred to this function here
    │ 
 40 │     global_config = p1;
    │     ─────────┬────────  
    │              ╰────────── Ownership of the value transferred to this place here
 41 │     global_config = p2;
    │     ─────────┬────────  
    │              ╰────────── Ownership handle lost without proper destruction
────╯
Error: Potential leak detected
    ╭─[preproccessed_file.c:1:1]
    │
 39 │     void* p2 = malloc(Foo2);
    │              ───────┬┬─────  
    │                     ╰──────── Ownership of the value transferred to this place here
    │                      │       
    │                      ╰─────── Ownership of the value transferred to this function here
    │ 
 41 │     global_config = p2;
    │     ─────────┬────────  
    │              ╰────────── Ownership of the value transferred to this place here
 42 │ }
    │ ┬  
    │ ╰── Function implicitly exited without proper destruction of that value
────╯